File: bin/qvalue_correction.R
#!/usr/bin/env Rscript

# Bradley Dec 2023

library(dplyr)
library(qvalue)
# Function to parse command-line arguments
parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  args_list <- list()
  i <- 1
  while (i <= length(args)) {
    if (args[i] %in% c("-f", "--file")) {
      args_list$file <- args[i + 1]
      i <- i + 1
    } else if (args[i] %in% c("-c", "--column")) {
      args_list$column <- as.numeric(args[i + 1])
      i <- i + 1
    } else if (args[i] %in% c("-n", "--new_q_val_column")) {
      args_list$new_q_val_column <- args[i + 1]
      i <- i + 1
    } else if (args[i] %in% c("-w", "--within_gene")) {
      args_list$within_gene <- args[i + 1]
      i <- i + 1
    }
    i <- i + 1
  }
  return(args_list)
}

# Parse the arguments
args <- parse_args()

# Check if all required arguments are provided
if (is.null(args$file) || is.null(args$column) || is.null(args$new_q_val_column) || is.null(args$within_gene)) {
  stop("All arguments -f, -c, -n, and -w are required")
}

# Assign arguments to variables
# file='/lustre/scratch123/hgi/teams/hgi/mo11/tmp_projects/saige/v1/work/66/6e28d628324ab5ddb2378689d8b838/output_Azimuth_predicted_celltype_l1__CD8_T__split_h5ad___1/nindep_100_ncell_100_lambda_2_tauIntraSample_0.5_cis_ENSG00000117862'
# column='13'
# new_column='qval'
# within=TRUE
file <- args$file
column <- args$column
new_column <- args$new_q_val_column
within <- args$within_gene

# Print the parsed arguments (optional, for debugging)
cat("File:", file, "\n")
cat("Column:", column, "\n")
cat("New Q-value column:", new_column, "\n")
cat("Within gene:", within, "\n")

# Load data
res = read.delim(file, header=F)
if(is.na(as.numeric(res[,as.numeric(column)][1]))){
    colnames(res) = res[1,]
    res = res[-1,]
}

# Extract p_values
p_values = as.numeric(res[,as.numeric(column)])

# Generate q-value object
qobj1 = qvalue(p = p_values); 
qobj = qobj1$qvalues

# qobj =p.adjust(p_values, method = c("BH"),n = length(p_values))
# qobj <- p.adjust(p_values, method = "BY", n = length(p_values))
# Add q-values
res[,new_column] = qobj

# Add local fdr
# res$lfdr = qobj$lfdr

# Replace the original file with the current one
write.table(res, file, col.names=T, quote=F, sep = "\t",row.names=F)

# Save a file with the variant with the minimum qvalue for this gene
min_file = gsub(".txt", "", file)
min_file = paste0(min_file, "_minimum_q.txt")
to_save = res[res[,new_column] == min(res[,new_column]),]
# If little significance, may get lots of results with the minimum qvalue
if(nrow(to_save > 1)){
    # If have lots of variants with same corrected value, take the one with the smallest original pvalue
    to_save=res[res[,column] == min(res[,column]),]
    # If have variants in complete LD - take the top
    if(nrow(to_save) > 1){
        to_save = to_save[1,]
    }
}

if(within == "TRUE"){
    write.table(to_save, min_file, col.names=T, quote=F, sep = "\t", row.names=F)
}
print("Performed q-value correction")

File: bin/qvalue_correction.py
#!/usr/bin/env python3

# Matiss Aug 2024

import argparse
import pandas as pd
from statsmodels.stats.multitest import multipletests
import qtl_qvalue
# Function to parse command-line arguments
def parse_args():
    parser = argparse.ArgumentParser(description="Process some arguments.")
    parser.add_argument("-f", "--file", required=True, help="Input file path")
    parser.add_argument("-c", "--column", type=int, required=True, help="Column number to extract p-values")
    parser.add_argument("-n", "--new_q_val_column", required=True, help="New column name for q-values")
    parser.add_argument("-w", "--within_gene", required=True, help="Flag to save the minimum q-value result within the gene")
    
    return parser.parse_args()

# Parse the arguments
args = parse_args()

# Assign arguments to variables
file = args.file
column = args.column - 1  # Python uses 0-based indexing
new_column = args.new_q_val_column
within_gene = args.within_gene

# Load data
res = pd.read_csv(file, sep='\t')
if pd.isna(pd.to_numeric(res.iloc[0, column], errors='coerce')):
    res.columns = res.iloc[0]
    res = res[1:]

# Extract p-values
p_values = pd.to_numeric(res.iloc[:, column], errors='coerce')

# Generate q-values using Benjamini-Hochberg method
# _, q_values, _, _ = multipletests(p_values, method='fdr_bh')
q_values = qtl_qvalue.QValue(p_values).qvalue()
# Add q-values to the DataFrame
res[new_column] = q_values

# Replace the original file with the current one
res.to_csv(file, sep='\t', index=False, header=True)

# Save a file with the variant with the minimum q-value for this gene
min_file = f"{file}_minimum_q.txt"
to_save = res[res[new_column] == res[new_column].min()]

# If little significance, may get lots of results with the minimum q-value
if len(to_save) > 1:
    # If have lots of variants with same corrected value, take the one with the smallest original p-value
    to_save = res[res.iloc[:, column] == res.iloc[:, column].min()]
    # If have variants in complete LD - take the top
    if len(to_save) > 1:
        to_save = to_save.iloc[0:1]

if within_gene == "TRUE":
    to_save.to_csv(min_file, sep='\t', index=False, header=True)

print("Performed q-value correction")
